modules/navier_stokes/src/executioners/SIMPLE.C

```diff
@@ -0,0 +1,1279 @@
+      for (auto system_i : index_range(_momentum_systems))
+        _console << " Momentum equation:"
+                 << (_momentum_systems.size() > 1
+                         ? std::string(" Component ") + std::to_string(system_i + 1) +
+                               std::string(" ")
+                         : std::string(" "))
+                 << COLOR_GREEN << ns_residuals[system_i] << COLOR_DEFAULT << std::endl;
+      _console << " Pressure equation: " << COLOR_GREEN << ns_residuals[momentum_residual.size()]
+               << COLOR_DEFAULT << std::endl;
+      if (_has_energy_system)
+      {
+        _console << " Energy equation: " << COLOR_GREEN
+                 << ns_residuals[momentum_residual.size() + 1] << COLOR_DEFAULT << std::endl;
+        if (_has_solid_energy_system)
+          _console << " Solid energy equation: " << COLOR_GREEN
+                   << ns_residuals[momentum_residual.size() + 2] << COLOR_DEFAULT << std::endl;
+      }
+    }
+
+    // Now we solve for the passive scalar equations, they should not influence the solution of the
+    // system above. The reason why we need more than one iteration is due to the matrix relaxation
+    // which can be used to stabilize the equations
+    if (_has_passive_scalar_systems)
+    {
+      _console << " Passive Scalar Iteration " << iteration_counter << std::endl;
+
+      // We set the options used by Petsc (preconditioners etc). We assume that every passive
+      // scalar equation uses the same options for now.
+      Moose::PetscSupport::petscSetOptions(_passive_scalar_petsc_options, solver_params);
+
+      iteration_counter = 0;
+      std::vector<Real> passive_scalar_residuals(_passive_scalar_systems.size(), 1.0);
+      while (iteration_counter < _num_iterations &&
+             !converged(passive_scalar_residuals, _passive_scalar_absolute_tolerance))
+      {
+        // We clear the caches in the passive scalar variables
+        for (auto system_i : index_range(_passive_scalar_systems))
+          _passive_scalar_systems[system_i]->residualSetup();
+
+        iteration_counter++;
+
+        // Solve the passive scalar equations
+        for (auto system_i : index_range(_passive_scalar_systems))
+          passive_scalar_residuals[system_i] =
+              solveAdvectedSystem(_passive_scalar_system_numbers[system_i],
+                                  *_passive_scalar_systems[system_i],
+                                  _passive_scalar_equation_relaxation[system_i],
+                                  _passive_scalar_linear_control,
+                                  _passive_scalar_l_abs_tol);
+
+        _console << "Iteration " << iteration_counter << " Initial residual norms:" << std::endl;
+        for (auto system_i : index_range(_passive_scalar_systems))
+          _console << _passive_scalar_systems[system_i]->name() << " " << COLOR_GREEN
+                   << passive_scalar_residuals[system_i] << COLOR_DEFAULT << std::endl;
+      }
+    }
+  }
```

Comment:
need a warning when you reach max non linear iterations

Reply:
I don't think so. The user can modify the number of iterations. Also, when you have a constant field, due to the cutom normalization of the residual norm you don't see convergence which can lead to these errors in the tests. In the future I can think about adding a different normalization, but for now when somebody has a constant field they will see no significant decrease in the residual.

Reply:
ok but I deleted 3 kernels and run and got no errors :)

Reply:
Yeah, this will be fixed when I merge the action and the executioner. Everything will be set up correctly there, for now we have to trust the users that they know what the Navier-Stokes equations look like.